src/streamed/state.ts

```diff
@@ -1,32 +1,443 @@
+      assert(tokenAddr, `missing token ${token}`)
+      return tokenAddr!
+    }
+  }
+
+  /**
+   * Gets the account data for the specified user, creating an empty one if it
+   * does not already exist.
+   */
+  private account(user: Address): Account {
+    let account = this.accounts.get(user)
+    if (account === undefined) {
+      account = {
+        balances: new Map(),
+        pendingWithdrawals: new Map(),
+        orders: [],
+      }
+      this.accounts.set(user, account)
+    }
+
+    return account
+  }
+
+  /**
+   * Updates a user's account balance.
+   */
+  private updateBalance(
+    user: Address,
+    token: TokenId | Address,
+    update: (amount: bigint) => bigint,
+  ): bigint {
+    const tokenAddr = this.tokenAddr(token)
+    const balances = this.account(user).balances
+    const newBalance = update(balances.get(tokenAddr) || BigInt(0))
+    balances.set(tokenAddr, newBalance)
+    return newBalance
+  }
+
+  /**
+   * Sets a pending withdrawal for a user for the specified token and amount.
+   */
+  private setPendingWithdrawal(
+    user: Address,
+    token: TokenId | Address,
+    batchId: number,
+    amount: bigint,
+  ): void {
+    const tokenAddr = this.tokenAddr(token)
+    if (this.options.strict) {
+      const existingBatch = this.account(user).pendingWithdrawals.get(tokenAddr)?.batchId
+      assert(
+        existingBatch === undefined || existingBatch === batchId,
+        `pending withdrawal for user ${user} modified from batch ${existingBatch} to ${batchId}`,
+      )
+    }
+    this.account(user).pendingWithdrawals.set(tokenAddr, { batchId, amount })
+  }
+
+  /**
+   * Clears a pending withdrawal.
+   */
+  private clearPendingWithdrawal(
+    user: Address,
+    token: TokenId | Address,
+  ): void {
+    const tokenAddr = this.tokenAddr(token)
+    this.account(user).pendingWithdrawals.delete(tokenAddr)
+  }
+
+  /**
+   * Adds a user order and returns the order ID of the newly created order.
+   */
+  private addOrder(user: Address, order: Order): number {
+    const userOrders = this.account(user).orders
+    userOrders.push(order)
+    return userOrders.length - 1
+  }
+
+  /**
+   * Retrieves an existing user order by user address and order ID.
+   *
+   * @throws If the order does not exist.
+   */
+  private order(user: Address, orderId: number): Order {
+    const order = this.account(user).orders[orderId]
+    assert(order, `attempted to retrieve missing order ${orderId} for user ${user}`)
+    return order
+  }
+
+  /**
+   * Updates a user's order's remaining amount. For unlimited orders, the
+   * remaining amount remains unchanged.
+   *
+   * @throws If the order does not exist.
+   */
+  private updateOrderRemainingAmount(
+    user: Address,
+    orderId: number,
+    update: (amount: bigint) => bigint,
+  ) {
+    const order = this.order(user, orderId)
+    if (
+      order.priceNumerator !== UNLIMITED_ORDER_AMOUNT &&
+      order.priceDenominator !== UNLIMITED_ORDER_AMOUNT
```

Comment:
> @fleupold This is another example of logic that probably already exists somewhere in dex-contracts.

I did not find this logic anywhere in `dex-contracts`.

Reply:
You are right. What a scandal! Let's create this as a utility method as soon as it's needed the second time.